sympy/combinatorics/rewritingsystem.py

```diff
@@ -139,6 +143,11 @@ def add_rule(self, w1, w2, check=False):
                 new = self.add_rule(w1, w2, check)
                 new_keys.update(new)
 
+        # Add new rules to the `new_rules` dictionary.
+        if self.reduction_automaton:
```

Comment:
Is this `if` statement necessary?

Reply:
I put it just to avoid the computation before the automaton is even constructed.

Reply:
But updating `new_rules` doesn't use the automaton so it doesn't matter if it's constructed or not. And don't you need to record `new_rules` during `_init_rules` to add them to the automaton that you constuct after?

(Also, this is another case where an underscore may be appropriate: `self._new_rules` because it's not an attribute that's meant to be used by the user).

Reply:
> And don't you need to record new_rules during _init_rules to add them to the automaton that you construct after? 

`self.rules` is used in the construction of the automaton, so, I've avoided the new_rules computation there. 

> Also, this is another case where an underscore may be appropriate: self._new_rules 

Yeah, I'll update it. Also, I think the `add_to_automaton` method should also be changed to `_add_to_automaton`.

Reply:
I see. Changing `add_to_automaton` to `_add_to_automaton` does make sense.